### PR TITLE
add a script for circuit compilation and verifier generation

### PR DESCRIPTION
**Pull request by Chengxuan**
Add a bash script to automate what's documented for circuit compilation, verification key generation and verifier library generation.

The `npm run gen` script by default processes up to 10 (configurable) circuits in parallel which hugely reduce the dev cycle.

### CODE REVIEW

File: solidity/contracts/lib/verifier_check_nullifiers.sol [hunk 1]
```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: GPL-3.0
```
Comment (jimthematrix) [inline solidity/contracts/lib/verifier_check_nullifiers.sol:1]: can the script be enhanced to skip certain circuit for the solidityverifier export step? the `check_nullifiers` circuit is only useful in off-chain use cases like a centralized component such as a notary to be able to validate nullifiers, in order to support anonymity in notarized transactions

File: zkp/generate.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
+
+# Check if PROVING_KEYS_ROOT is not set or is empty
+if [ -z "$PROVING_KEYS_ROOT" ]; then
+  echo "Error: PROVING_KEYS_ROOT is not set."
+  exit 1
+fi
+
+# Define arrays
+X_values=(
```
Comment (jimthematrix) [inline zkp/generate.sh:10]: maybe give it a better name like `circuit_names`

File: zkp/generate.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    "anon_nullifier"
+    "anon_enc_nullifier"
+    "nf_anon"
+    "nf_anon_nullifier"
+    "check_hashes_value"
+    "check_inputs_outputs_value"
+    "check_nullifier_value"
+    "check_nullifiers"
+)
+Y_values=(
```
Comment (jimthematrix) [inline zkp/generate.sh:22]: maybe give it a better name like `ptau_names`

File: zkp/generate.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+fi
+
+# Loop through the arrays
+for i in "${!X_values[@]}"; do
+    X=${X_values[$i]}
+    Y=${Y_values[$i]}
+    echo "Processing $X"
+
+    CIRCOM_INPUT="circuits/$X.circom"
+    PTIAU_FILE="$HOME/Downloads/$Y.ptau"
```
Comment (jimthematrix) [inline zkp/generate.sh:48]: can we add a step to download these files as well if not already there?

**Review by jimthematrix — CHANGES_REQUESTED** [review]
Looks great @Chengxuan! something that I've been wanting to do as well. will be super useful!

File: solidity/contracts/lib/verifier_check_nullifiers.sol (hunk 1 above)
Comment (Chengxuan) [reply solidity/contracts/lib/verifier_check_nullifiers.sol:1]: I think maybe this one can be a follow-on PR, we need to consider whether the script should be smart enough to be aware of "profiles". I don't want a bash script to get too complex. Maybe we should switch to a js script at that stage.

File: zkp/generate.sh [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    circom "$CIRCOM_INPUT" --output "$PROVING_KEYS_ROOT" --r1cs
+
+    echo "Generate test proving key for $X with $Y"
+    snarkjs groth16 setup "$PROVING_KEYS_ROOT/$X.r1cs" "$PTIAU_FILE" "$ZKEY_OUTPUT"
+    
+    echo "Generate verification key for $X"
+    snarkjs zkey export verificationkey "$PROVING_KEYS_ROOT/$X.zkey" "$PROVING_KEYS_ROOT/$X.-vkey.json"
+    
+    echo "Generate solidity verifier for $X"
+    snarkjs zkey export solidityverifier "$PROVING_KEYS_ROOT/$X.zkey" "../solidity/contracts/lib/verifier_$X.sol"
```
[the remark this answers is not in the document]
Comment (Chengxuan) [reply zkp/generate.sh]: @jimthematrix I raised a separate PR for this: https://github.com/Chengxuan/zeto/pull/1 as it touched quite a few files

File: solidity/contracts/lib/verifier_check_nullifiers.sol (hunk 1 above)
Comment (Chengxuan) [reply solidity/contracts/lib/verifier_check_nullifiers.sol:1]: coverted the script to use JS

**Comment by Chengxuan** [comment]
@jimthematrix thanks for the review comments. Switched the script to an NPM script with all comment addressed

**Review by jimthematrix — APPROVED** [review]

**Pull request merged**